glib/src/subclass/signal.rs

```diff
@@ -592,27 +595,31 @@ impl Signal {
         let return_type = self.return_type;
 
         let class_handler = class_handler.map(|class_handler| {
-            Closure::new(move |values| unsafe {
-                let instance = gobject_ffi::g_value_get_object(values[0].to_glib_none().0);
-                let res = class_handler(&SignalClassHandlerToken(instance as *mut _, return_type.into(), values.as_ptr()), values);
-
-                if return_type == Type::UNIT {
-                    if let Some(ref v) = res {
-                        panic!("Signal has no return value but class handler returned a value of type {}", v.type_());
-                    }
-                } else {
-                    match res {
-                        None => {
-                            panic!("Signal has a return value but class handler returned none");
-                        }
-                        Some(ref v) => {
-                            assert!(v.type_().is_a(return_type.into()), "Signal has a return type of {} but class handler returned {}", Type::from(return_type), v.type_());
+            match class_handler {
+                SignalClassHandler::Function(class_handler) => {
+                    Closure::new(move |values| {
+                        let res = class_handler(values);
+
+                        if return_type == Type::UNIT {
+                            if let Some(ref v) = res {
+                                panic!("Signal has no return value but class handler returned a value of type {}", v.type_());
+                            }
+                        } else {
+                            match res {
+                                None => {
+                                    panic!("Signal has a return value but class handler returned none");
+                                }
+                                Some(ref v) => {
+                                    assert!(v.type_().is_a(return_type.into()), "Signal has a return type of {} but class handler returned {}", Type::from(return_type), v.type_());
+                                }
+                            }
                         }
-                    }
-                }
 
-                res
-            })
+                        res
+                    })
+                },
+                SignalClassHandler::Closure(closure) => closure.as_ref().clone(),
```

Comment:
Where does this check for a valid return type now?

Reply:
I just took out the `class_closure` bit, I don't think it's possible to do that safely and also do a check for the return type